packages/cli/src/commands/hydrogen/env/push__unstable.ts

```diff
@@ -0,0 +1,243 @@
+      renderWarning({
+        headline: 'Environment not found',
+        body: `We could not find an environment matching the name '${environment}'.`
+      });
+      process.exit(1);
+    } else if (matchedEnvironments.length === 1) {
+      const {id, name, branch, type} = matchedEnvironments[0] ?? {};
+      validatedEnvironment = {id, name, branch, type};
+    } else {
+      // Prompt the user for a selection if there are multiple matches
+      const selection = await renderSelectPrompt({
+        message: `There were multiple environments found with the name ${environment}:`,
+        choices:
+        [
+          ...matchedEnvironments.map(({id, name, branch, type, url}) => ({
+            label: `${name} (${branch}) ${type} ${url}`,
+            value: id,
+          })),
+        ]
+      });
+      const {id, name, branch, type} = matchedEnvironments.find(({id}) => id === selection) ?? {};
+      validatedEnvironment = {id, name, branch, type};
+    }
+  } else {
+    // Environment flag not passed
+    const choices = [
+      ...environments.map(({id, name, branch}) => ({
+        label: branch ? `${name} (${branch})` : name,
+        value: id,
+      })),
+    ];
+
+    const pushToBranchSelection = await renderSelectPrompt({
+      message: 'Select a set of environment variables to overwrite:',
+      choices,
+    });
+
+    const {id, name, branch, type} = environments.find(({id}) => id === pushToBranchSelection) ?? {};
+    validatedEnvironment = {id, name, branch, type};
+  }
+
+  // Fetch remote variables
+  const {environmentVariables = []} = await getStorefrontEnvVariables(
+    session,
+    config.storefront.id,
+    validatedEnvironment.branch ?? undefined,
+  ) ?? {};
+
+  // Normalize variables
+  const remoteVars = environmentVariables.filter(({isSecret, readOnly}) => !isSecret && !readOnly);
+  const compareableRemoteVars = remoteVars
+    .filter(({isSecret, readOnly}) => !isSecret && !readOnly)
+    .map(({key, value}) => `${key}=${value}`).join('\n') + '\n';
+
+  const compareableLocalVars = Object.keys(localVariables).reduce((acc, key) => {
+    const {isSecret, readOnly} = environmentVariables.find((variable) => variable.key === key) ?? {};
+    if (isSecret || readOnly) return acc;
+    return [...acc, `${key}=${localVariables[key]}`];
+  }, [] as string[]).join('\n') + '\n';
+
+  // Confirm changes and show a generate diff of changes
+  if (!validatedEnvironment.name) process.exit(1);
+
+  // Generate a list of remote read only or secrets
+  const remoteReadOnlyOrSecrets = environmentVariables.reduce(
+    (acc, {isSecret, readOnly, key}) => {
+      if (!isSecret && !readOnly) return acc;
+      const localVar = localVariables[key];
+      const remoteVar = environmentVariables.find((variable) => variable.key === key);
+      if (localVar === remoteVar?.value) return acc;
+      return [...acc, key];
+    },
+    [] as string[],
+  );
+
+  if (compareableLocalVars === compareableRemoteVars) {
```

Comment:
I might have missed something but, shouldn't the variables (remote and local) be sorted in a specific way before building this thing and comparing them?

Or perhaps the comparison should happen checking variables 1 by 1 instead of building strings where order can matter? 🤔

Reply:
Hm, good point actually. It's funny then how it's just been working fine with real data! Luck of the draw maybe. I will add this for sure.

Reply:
Added! https://github.com/Shopify/hydrogen/compare/01a3b6580e497131993a9287083f1afba89bd105..8493852bb19ecd257063dea73c2507fe87b5c161